### PR TITLE
#457 add validation to maintain uniqueness of product and variations …

### PR DESCRIPTION
**Pull request by lnghia**
…on sku, gtin, slug separately

### CODE REVIEW

**Review by nguyenvanhadncntt — APPROVED** [review]

File: product/src/main/java/com/yas/product/exception/ApiExceptionHandler.java [hunk 1]
```diff
@@ -75,6 +75,23 @@ protected ResponseEntity<Object> handleDuplicatedName(DuplicatedNameException e)
     return ResponseEntity.badRequest().body(errorVm);
   }
 
+  @ExceptionHandler(DuplicatedSlugException.class)
```
Comment (toannguyenk) [inline product/src/main/java/com/yas/product/exception/ApiExceptionHandler.java:78]: why dont we reuse DuplicatedNameException ?

**Review by toannguyenk — APPROVED** [review]

File: product/src/main/java/com/yas/product/service/ProductService.java [hunk 2]
```diff
@@ -1,7 +1,6 @@
 package com.yas.product.service;
 
-import com.yas.product.exception.BadRequestException;
```
Comment (duccaominh) [inline product/src/main/java/com/yas/product/service/ProductService.java:3]: Should not using wild card `*` import

File: product/src/main/java/com/yas/product/service/ProductService.java [hunk 3]
```diff
@@ -83,7 +82,38 @@ public ProductListGetVm getProductsWithFilter(int pageNo, int pageSize, String p
         );
     }
 
+    private boolean isProductWithSlugAvailable(String slug) {
+        return productRepository.findBySlugAndIsActiveTrue(slug).isPresent();
```
Comment (duccaominh) [inline product/src/main/java/com/yas/product/service/ProductService.java:86]: refactor to `existsBy`, no need to call .isPresent()

File: product/src/main/java/com/yas/product/service/ProductService.java [hunk 4]
```diff
@@ -83,7 +82,38 @@ public ProductListGetVm getProductsWithFilter(int pageNo, int pageSize, String p
...
+    }
+
+    private boolean isProductWithSkuAvailable(String sku) {
+        return productRepository.findBySkuAndIsActiveTrue(sku).isPresent();
```
Comment (duccaominh) [inline product/src/main/java/com/yas/product/service/ProductService.java:90]: same above

File: product/src/main/java/com/yas/product/service/ProductService.java [hunk 5]
```diff
@@ -83,7 +82,38 @@ public ProductListGetVm getProductsWithFilter(int pageNo, int pageSize, String p
...
+    }
+
+    private boolean isProductWithGtinAvailable(String gtin) {
+        return productRepository.findByGtinAndIsActiveTrue(gtin).isPresent();
```
Comment (duccaominh) [inline product/src/main/java/com/yas/product/service/ProductService.java:94]: same above

**Pull request merged**